Format Smart

W3stside opened the pull request:
Formats amounts in a "smart" way lol

Could be my most ironic PR ever if the logic isn't smart :roll_eyes: 

Anyways why I like this more than the last one on dex-react:
1. no dependecy on `Intl.NumberFormat` library, which isn't supported by all browsers
2. no `parseFloat` shenanigans aka no floating numbers
3. I made it, therefor i like it

Updated tests, all green, can also be manualy tested inside node if you build via:
```bash
yarn build:watch
# then inside some new terminal window in the project root

node
```
```js
const utils = require('.')
utils.formatSmart(utils.toBN('125492310294123123'))
// '125,492.31T'
utils.formatSmart(utils.toBN('125492310294123'))
// '125.492T'
utils.formatSmart(utils.toBN('1254923102941'))
// '1.254T'
utils.formatSmart(utils.toBN('125492310294'))
// '125.492B'
utils.formatSmart(utils.toBN('1254923102'))
// '1.254B'
utils.formatSmart(utils.toBN('125492302'))
// '125,492,302'
utils.formatSmart(utils.parseAmount('125492302.1235123'))
// '126,727,425'
utils.formatSmart(utils.parseAmount('125492302.1235123', 18), 18)
// '125,492,302.1'
```

W3stside committed:
Function: smartFormat

W3stside committed:
tests

W3stside committed:
lock

W3stside committed:
Tests: renamed file + fixed tests

W3stside committed:
Constants: added some, moved some

1. added BN constants used in smart

W3stside committed:
Format: changed core formatSmart logic

1. removed Intl library dependency
2. optimised as much as possible ;)

alfetopito commented on src/const.ts:
```diff
@@ -1,11 +1,18 @@
 export const ONE = new BN(1)
 export const TWO = new BN(2)
 export const TEN = new BN(10)
+export const ONE_HUNDRED_K = new BN('100000')
+export const ONE_MILLION_BN = new BN('1000000')
+export const TEN_MILLION = new BN('10000000')
+export const ONE_BILLION_BN = new BN('1000000000')
+export const ONE_TRILLION_BN = new BN('1000000000000')
```
There are 3 different suffixes: `_K`, `_BN` and no suffix

Can you pick one? :D

W3stside replied:
:sweat_smile: NO

alfetopito commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  let finalPrecision: number
+
+  // normal format + separator
+  if (integerPart.lt(ONE_HUNDRED_K)) {
+    // 100,000.3424
```
The code doesn't match the comment.
100'000 not less than 100'000.
is this a bad example or should be less than equal?

W3stside replied:
oops ur right, fixing

alfetopito commented on test/utils/format/formatSmart.spec.ts:
```diff
@@ -0,0 +1,184 @@
+
+  // TODO: Define what for arbitrarily big amounts
+  test('uint max value', async () => {
+    const expected = '115,792,089,237,316,195,423,570,985,008,687,907,853,269,984,665.64T'
```
May I suggest we return this instead?

```suggestion
    const expected = 'A very big number. Trust me, it\'s THAT big.'
```

W3stside replied:
that amount is my ETH goals

alfetopito reviewed (approved):
I like it!

W3stside committed:
Fixed constant names (consistency) + comment

Velenir commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  // no relevant decimal section
+  if (decimalPart.isZero()) return formattedInteger
+
+  const formattedDecimal = decimalPart
+    .toString()
+    .padStart(precision, '0') // Pad the decimal part with leading zeros
+    .replace(/0+$/, '') // Remove the right zeros
```
Can `decimalPart` be `0`? Then `.replace(/0+$/, '')` will produce an empty string

W3stside replied:
not sure

W3stside replied:
oops this responded to wrong comment, it is fine as there is logic to catch 0 decimals in other places (just ignore it and it will go away)

anxolin commented on src/const.ts:
```diff
@@ -15,8 +22,18 @@ export const ALLOWANCE_MAX_VALUE = TWO.pow(new BN(256)).sub(ONE) // 115792089237
 export const ALLOWANCE_FOR_ENABLED_TOKEN = TWO.pow(new BN(128)) // 340282366920938463463374607431768211456
 
 // Default formatting constants
+export const ELLIPSIS = '...'
```
Why do we use the ellipsis? 
I don't see it being used in this PR.

Also, are you sure we want the ellipsis as 3 dots? There's a UNICODE for that https://www.compart.com/en/unicode/U+2026

W3stside replied:
not sure i was just moving it, didnt add it. I think it may be used in dex-react?

W3stside replied:
it's used inside:
```ts
export function abbreviateString(inputString: string, prefixLength: number, suffixLength = 0): string {
  // abbreviate only if it makes sense, and make sure ellipsis fits into word
  // 1. always add ellipsis
  // 2. do not shorten words in case ellipsis will make the word longer
  // 3. min prefix == 1
  // 4. add suffix if requested
  const _prefixLength = Math.max(1, prefixLength)
  if (inputString.length < _prefixLength + ELLIPSIS.length + suffixLength) {
    return inputString
  }
  const prefix = inputString.slice(0, _prefixLength)
  const suffix = suffixLength > 0 ? inputString.slice(-suffixLength) : ''
  return prefix + ELLIPSIS + suffix
}
```

W3stside replied:
again tho i just moved it, not using it specifically in my PR

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  let precision: number
+
+  let decimals = DEFAULT_DECIMALS
+  const smallLimit = DEFAULT_SMALL_LIMIT_AS_WEI
```
I wouldn't do this rename of the constant DEFAULT_SMALL_LIMIT_AS_WEI

W3stside replied:
yep

W3stside replied:
this needed to be change as it has to be parameterised

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  let decimals = DEFAULT_DECIMALS
+  const smallLimit = DEFAULT_SMALL_LIMIT_AS_WEI
+
+  if (!params || ('amount' in params && !params.amount)) {
```
If this is the only case that returns right away. I would make it a standalone IF without ELSEs as the first line in the function.

W3stside replied:
done

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  if (amount.isZero()) return amount.toString()
+
+  const thousandSymbol = THOUSANDS_SYMBOL
+  const decimalSymbol = DECIMALS_SYMBOL
```
why do we do this rename of the consts?

W3stside replied:
not sure, I was reusing the `formatAmount` logic as a base. I think it's because of the overloading

W3stside replied:
yup, not needed! good catch

anxolin commented on src/const.ts:
```diff
@@ -15,8 +22,18 @@ export const ALLOWANCE_MAX_VALUE = TWO.pow(new BN(256)).sub(ONE) // 115792089237
+// Represented as WEI to use BN comparison
+// Why? Would not be possible otherwise, as
+// BN does not accept fractional values
+export const DEFAULT_SMALL_LIMIT_AS_WEI = new BN(toWei('0.001'))
```
BigNumbers supports that decimal values, not sure what's best here

W3stside replied:
ah shit that's true, mm idk

W3stside replied:
changed it to BigNumber, makes more sense and easier to read

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+interface CompactionOptions {
+  precision: number,
+  thousandsSymbol: string,
+  decimalSymbol: string
```
Those 2 symbols, do they really need to be part of the parameters? they are known consts

W3stside replied:
good catch

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  }: CompactionOptions): string {
+  // e.g TRILLION_123_123.div(ONE_TRILLION) = 123123.123123123
+  const numAsFraction = baseBN.mul(TEN.pow(new BN(precision))).div(representationBN)
+  const { integerPart, decimalPart } = _decomposeBn(numAsFraction, DEFAULT_LARGE_NUMBER_PRECISION, DEFAULT_LARGE_NUMBER_PRECISION)
```
Why do you need to pass known constants to a private function?

W3stside replied:
because i use _decomposeBn in different places and their precisions are different

W3stside replied:
nope ur right changing

anxolin commented on src/const.ts:
```diff
@@ -15,8 +22,18 @@ export const ALLOWANCE_MAX_VALUE = TWO.pow(new BN(256)).sub(ONE) // 115792089237
+// BN does not accept fractional values
+export const DEFAULT_SMALL_LIMIT_AS_WEI = new BN(toWei('0.001'))
+export const DEFAULT_THOUSANDS_SYMBOL = ','
+export const DEFAULT_DECIMALS_SYMBOL = '.'
```
We did have already this consts, why we need them again hardcoded?
Is this coherent with the other formatters?

W3stside replied:
yeah i just moved them into `const`

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  decimalSymbol: string
+}
+
+function _compactLargeBNNumberToString(
```
I'm a bit confused by this name.
Also, nitpick. For camel case spelling, I think is better that for acronyms like BN we spell it like Bn --> I think BnNumber is more readable than BNNumber (what we try to follow in other parts.

W3stside replied:
good idea, changing and made name more specific.

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+
+function _compactLargeBNNumberToString(
+  baseBN: BN,
+  representationBN: BN,
```
Do we need to include the work "BN" in the name of the var?
Also. I was confused by `representationBN` I had to look the implementation to understand. 
I cannot come up with a better name than `baseUnitsPerRepresentationUnits` (I feel is too long, but at least I feel is more clear). Maybe you can come up with a better name?

W3stside replied:
i like that name

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+    decimalSymbol,
+  }: CompactionOptions): string {
+  // e.g TRILLION_123_123.div(ONE_TRILLION) = 123123.123123123
+  const numAsFraction = baseBN.mul(TEN.pow(new BN(precision))).div(representationBN)
```
Note: Fractions are different as decimals. Decimal can be expressed as a fraction, but they are 2 different things.

Also, BN doesn't allow decimals. Every time we work with decimals, we would be better off with BigNumber instead of BN.

W3stside replied:
true, changing

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  if (decimalPart.isZero()) return formattedInteger
+
+  const formattedDecimal = decimalPart
+    .toString()
```
Add a `. toString(10)` here, or it may add exponentiation for really big numbers --> I think...

W3stside replied:
ok done

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  const formattedDecimal = decimalPart
+    .toString()
+    .padStart(precision, '0') // Pad the decimal part with leading zeros
+    .replace(/0+$/, '') // Remove the right zeros
```
If we are using this in the order book, should't be this part parametrized?  (so for the order book we can leave the zeros on the right)

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  precision: number = DEFAULT_LARGE_NUMBER_PRECISION,
+): string {
+  // Is fraction
+  if (integerPart.isZero()) {
```
I think the comment above is misleading. 
I understand there you are only testing that the value is `<1`, nothing related to being a fraction

3/2 is a fraction with value 1.5

W3stside replied:
yeah true, changed

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
-const DEFAULT_THOUSANDS_SYMBOL = ','
-const DEFAULT_DECIMALS_SYMBOL = '.'
+
+const _formatDecimalsFromWeiForDisplay = (decimalsToConvert: BN, decimalSymbol: string = DEFAULT_DECIMALS_SYMBOL) => '0' + decimalSymbol + fromWei(decimalsToConvert).slice(2)
```
I think this one-line expressions are meant for simple definitions 
This one is a bit hard to read (difficult name, then types, and theres's sth odd in that fromWei/slice. 
Can you make it a normal function?

W3stside replied:
done

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  // if decimals are zero
+  if (decimalPart.isZero()) return _formatNumber(integerPart.toString())
+
+  let finalPrecision: number
```
🤘I like this part. The comments are also helpfull

W3stside replied:
thanks!

anxolin commented on src/utils/format.ts:
```diff
@@ -15,27 +28,170 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  const amountRaw = amount.divRound(TEN.pow(new BN(amountPrecision - decimals)))
+  const integerPart = amountRaw.div(TEN.pow(new BN(decimals))) // 165000 / 10000 = 16
+  const decimalPart = amountRaw.mod(TEN.pow(new BN(decimals))) // 165000 % 10000 = 5000
+  const decimalsPadded = decimalPart.toString().padStart(decimals, '0')
```
why duplicating this infor?

Also, I think every `toString()` should have the `toString(10)` parameter?

W3stside replied:
i thought base 10 was default? can add

W3stside replied:
I am adding it here as i use it in several places

anxolin replied:
I'm not sure if it's only for BigNumber, but I remember some issues we had regarding this, cause it was using the scientific notation for large numbers

W3stside replied:
the native `.toString()` uses `10` as default but maybe `BN`/`BigNumber` doesn't. Makes sense to keep it consistent though

alfetopito replied:
It does use base 10 by default, but when base is not passed as parameter, it defaults to exponential notation depending on the precision https://mikemcl.github.io/bignumber.js/#toS

anxolin commented on test/utils/format/formatSmart.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    expect(formatSmart({ amount, precision: 6, decimals: 2 })).toEqual('4,567,890.12')
+  })
+
+  test('4,567,890.125555 units, round up', async () => {
```
The test is good. But probably you would want to change the test description to --> `4,567,890.125555 units, it doesn't round up`

W3stside replied:
it does round up, from `4,567,890.12555` to `4,567,890.13`

W3stside replied:
values under `10m` keep 2 decimals and round, according to the Gnosis Safe spec

anxolin commented on test/utils/format/formatSmart.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    expect(formatSmart({ amount, precision: 6, decimals: 0 })).toEqual('4,567,890')
+  })
+
+  test('4,567,890.125555 units, round up', async () => {
```
same as my comment ☝️

anxolin commented on test/utils/format/formatSmart.spec.ts:
```diff
@@ -0,0 +1,184 @@
+})
+
+describe('Big amounts', () => {
+  // TODO: Considering showing K,M,B,...
```
Can you make this comment a issue if it really make sense?

W3stside replied:
this should have been removed

anxolin commented on test/utils/format/formatSmart.spec.ts:
```diff
@@ -0,0 +1,184 @@
+    expect(formatSmart(new BN(toWei(new BN('1000000000'), 'ether')), DEFAULT_PRECISION)).toEqual('1B')
+  })
+
+  // TODO: Define what for arbitrarily big amounts
```
I would say, at some point we show in scientific `1.15792089 * 10**20 DAI` (using markup?? --> https://developer.mozilla.org/en-US/docs/Web/HTML/Element/sup)

anxolin reviewed (commented):
Great job, it improves a lot the last PR!
Specially liked the tests.

I added many comments, most of it are nitpicks

W3stside committed:
test comments

W3stside committed:
fraction => decimals & pass constants in

W3stside committed:
Many PR review fixes

1. pass constants directly into fns when necessary
2. remove unused params
3. remove unsused options
4. change some if/else statements to use single if, when necessary
5. explicit .toString(10)

W3stside committed:
handle division math with BigNumber

1. dont use BN and precision math

W3stside committed:
change DEFAULT_SMALL_LIMIT to string

1. let fn handle conversion to BigNumber

alfetopito commented on src/utils/format.ts:
```diff
@@ -15,27 +27,159 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+function _formatDecimalsForDisplay(decimalsToConvert: BigNumber, decimalSymbol: string = DEFAULT_DECIMALS_SYMBOL) {
+  const prefix = '0'
+  // e.g 00012366123
+  const decimalsWithoutIntegerOrSymbol = decimalsToConvert.toString(10).slice(2)
```
Got confused at first.
This assumes `decimalsToConvert` will never have an integer part.

Maybe to be sure would be good to subtract any integer? https://mikemcl.github.io/bignumber.js/#int

```suggestion
  const decimalsWithoutIntegerOrSymbol = (decimalsToConvert.minus(decimalsToConvert.integerValue())).toString(10).slice(2)
```

W3stside replied:
true, wasn't really accounting for negative numbers

W3stside replied:
what is `.integerValue()`?

W3stside replied:
nvm clicked the link like i probably should have done in the first place and am G2G

alfetopito commented on src/utils/format.ts:
```diff
@@ -15,27 +27,159 @@ function _getLocaleSymbols(): { thousands: string; decimals: string } {
+  }
+
+  // amount is already zero
+  if (amount.isZero()) return amount.toString(10)
```
The _Needless Optimization Group_ reviewed you code and made the following suggestion:

Let's save 1 function call!

```suggestion
  if (amount.isZero()) return '0'
```
/shrug

W3stside replied:
\o/ great idea

W3stside committed:
if ZERO return '0'

Co-authored-by: Leandro Boscariol <alfetopito@users.noreply.github.com>

W3stside committed:
PR fix concerning decimal/integer format
